File: worker/src/deelnemen.js
// Het deelnameformulier lezen, invullen en versturen.
// Zelfde regels als deelnemen.py: nooit gokken bij twijfel, geen lege verplichte
// velden versturen, en bij een captcha stoppen we — dan komt het op de digest.
import { UA, haal, ontsnap, stripHtml } from "./bronnen.js";
import { budget } from "./budget.js";
import { schatSchifting, soortVraag, zoekAntwoord } from "./antwoord.js";

const CAPTCHA_SPOREN = ["recaptcha", "g-recaptcha", "hcaptcha", "h-captcha",
                        "cf-turnstile", "turnstile", "friendlycaptcha", "captcha"];
const LOGIN_SPOREN = ["wachtwoord", "password", "inloggen om deel te nemen", "log in om"];
const GELUKT = ["bedankt", "je deelname", "uw deelname", "deelname geregistreerd",
                "succesvol", "gelukt", "bevestigingsmail", "thank you", "veel succes"];

const VELDPATRONEN = [
  [/voornaam|firstname|first_name|fname|given/, "voornaam"],
  [/achternaam|lastname|last_name|lname|surname|familienaam/, "achternaam"],
  [/\bnaam\b|fullname|full_name|\bname\b/, "volledigeNaam"],
  [/e.?mail/, "email"],
  [/telefoon|gsm|mobiel|phone|tel\b/, "telefoon"],
  [/straat|street|adres|address(?!.*mail)/, "straat"],
  [/huisnummer|nummer|house.?number|streetnumber/, "huisnummer"],
  [/postcode|zip|postal/, "postcode"],
  [/gemeente|woonplaats|stad|city|plaats/, "gemeente"],
  [/land|country/, "land"],
  [/geboorte|birth|dob/, "geboortedatum"],
];

// Overzichtssites tonen een detailpagina met een knop naar de echte wedstrijd.
// Die hop misten we: we zochten een formulier op een pagina die er geen heeft.
const DOORKLIK = /deelnem|meedoen|doe\s*mee|naar de (wedstrijd|actie)|win nu|inschrijv|naar de site|bezoek/i;

export function zoekDoorklik(html, basisUrl) {
  const huidig = new URL(basisUrl);
  const eigen = huidig.hostname.replace(/^www\./, "");
  const kandidaten = [];

  for (const [, href, binnenkant] of html.matchAll(/<a\b[^>]*href=["']([^"']+)["'][^>]*>([\s\S]*?)<\/a>/gi)) {
    const tekst = stripHtml(binnenkant);
    if (!DOORKLIK.test(tekst)) continue;

    let doel;
    try {
      doel = new URL(ontsnap(href), basisUrl);
    } catch {
      continue;
    }
    if (!/^https?:$/.test(doel.protocol)) continue;
    if (doel.toString().split("#")[0] === basisUrl.split("#")[0]) continue;

    const extern = doel.hostname.replace(/^www\./, "") !== eigen;
    // Een link terug naar de homepage van de overzichtssite is geen wedstrijd.
    if (!extern && doel.pathname.replace(/\/+$/, "") === "") continue;

    kandidaten.push({ url: doel.toString(), extern });
  }

  // Extern heeft voorrang (dat is meestal het merk zelf), maar een doorklik
  // binnen dezelfde site telt ook: sommige overzichtssites hosten het
  // formulier gewoon zelf.
  const gekozen = kandidaten.find((k) => k.extern) || kandidaten[0];
  return gekozen ? gekozen.url : null;
}

export function bevatCaptcha(html) {
  const laag = html.toLowerCase();
  return CAPTCHA_SPOREN.some((spoor) => laag.includes(spoor));
}

function kenmerken(tag) {
  const uit = {};
  for (const [, naam, waarde] of tag.matchAll(/([a-zA-Z-]+)(?:=["']([^"']*)["'])?/g)) {
    uit[naam.toLowerCase()] = waarde ?? "";
  }
  return uit;
}

export function leesFormulieren(html) {
  const labels = {};
  const labelTags = html.matchAll(/<label[^>]*for=["']([^"']+)["'][^>]*>([\s\S]*?)<\/label>/gi);
  for (const [, voor, tekst] of labelTags) labels[voor] = stripHtml(tekst);

  const formulieren = [];
  for (const [blok, opening] of html.matchAll(/(<form\b[^>]*>)[\s\S]*?<\/form>/gi)) {
    const kop = kenmerken(opening);
    const velden = [];
    // Keuzes per veldnaam: nodig om een wedstrijdvraag te kunnen beantwoorden.
    const keuzes = {};

    // <select name="antwoord"><option value="a">Brussel</option>...
    for (const [blokje, selectTag] of blok.matchAll(/(<select\b[^>]*>[\s\S]*?<\/select>)/gi)) {
      const naam = kenmerken(selectTag).name;
      if (!naam) continue;
      for (const [, waarde, tekst] of blokje.matchAll(/<option[^>]*value=["']([^"']*)["'][^>]*>([\s\S]*?)<\/option>/gi)) {
        if (!waarde) continue;
        (keuzes[naam] ||= []).push({ value: waarde, tekst: stripHtml(tekst) });
      }
    }
    for (const [, soort, rest] of blok.matchAll(/<(input|select|textarea)\b([^>]*)>/gi)) {
      const k = kenmerken(rest);
      velden.push({
        type: (k.type || (soort.toLowerCase() === "input" ? "text" : soort.toLowerCase())).toLowerCase(),
        name: k.name || "",
        id: k.id || "",
        value: ontsnap(k.value || ""),
        placeholder: k.placeholder || "",
        required: "required" in k,
        checked: "checked" in k,
      });
    }
    // Radiokeuzes: de tekst staat in het <label for="...">.
    for (const veld of velden) {
      if (veld.type !== "radio" || !veld.name) continue;
      (keuzes[veld.name] ||= []).push({
        value: veld.value || "on",
        tekst: labels[veld.id] || veld.value || "",
      });
    }

    formulieren.push({
      action: ontsnap(kop.action || ""),
      method: (kop.method || "get").toLowerCase(),
      velden,
      keuzes,
    });
  }
  return { formulieren, labels };
}

export function kiesFormulier(formulieren) {
  let beste = null;
  for (const formulier of formulieren) {
    const heeftEmail = formulier.velden.some((v) => v.type === "email" || /e.?mail/i.test(v.name));
    const zichtbaar = formulier.velden.filter(
      (v) => !["hidden", "submit", "button", "image"].includes(v.type));
    if (!heeftEmail || !zichtbaar.length) continue;
    if (!beste || zichtbaar.length > beste.aantal) beste = { formulier, aantal: zichtbaar.length };
  }
  return beste ? beste.formulier : null;
}

function raadVeld(veld, labels) {
  if (veld.type === "email") return "email";
  if (veld.type === "tel") return "telefoon";
  const aanwijzing = [veld.name, veld.id, veld.placeholder, labels[veld.id] || ""]
    .join(" ").toLowerCase();
  for (const [patroon, sleutel] of VELDPATRONEN) if (patroon.test(aanwijzing)) return sleutel;
  return null;
}

export function bouwPayload(formulier, labels, deelnemer, context = {}) {
  const payload = new URLSearchParams();
  const onbekend = [];
  const uitleg = [];
  const keuzes = formulier.keuzes || {};

  // Een wedstrijd- of schiftingsvraag invullen op basis van wat er op de
  // pagina staat. Lukt dat niet overtuigend, dan blijft het veld onbekend en
  // gaat de wedstrijd naar "zelf doen" — liever dat dan een gok die je
  // deelname toch ongeldig maakt.
  const probeerVraag = (veld, aanwijzing) => {
    const soort = soortVraag(aanwijzing);
    if (!soort) return false;
    const gevonden = soort === "schifting"
      ? schatSchifting(aanwijzing, context.pagina || "", context.context || "")
      : zoekAntwoord(aanwijzing, keuzes[veld.name] || [], context.pagina || "", context.context || "");
    if (!gevonden) return false;
    payload.set(veld.name, gevonden.waarde);
    uitleg.push(`${soort === "schifting" ? "schifting" : "wedstrijdvraag"}: ` +
                `"${gevonden.waarde}" — ${gevonden.reden}`);
    return true;
  };

  for (const veld of formulier.velden) {
    const naam = veld.name;

    if (["file", "password"].includes(veld.type)) {
      onbekend.push(`${veld.type}-veld (${naam || veld.id})`);
      continue;
    }
    if (!naam || ["submit", "button", "image"].includes(veld.type)) continue;

    if (veld.type === "hidden") {
      payload.set(naam, veld.value);
      continue;
    }

    if (veld.type === "checkbox") {
      const label = `${labels[veld.id] || ""} ${naam}`.toLowerCase();
      if (/voorwaard|reglement|akkoord|privacy|18|leeftijd|terms/.test(label)) {
        payload.set(naam, veld.value || "on");
      } else if (/nieuwsbrief|newsletter|aanbieding|partner|marketing|mailing/.test(label)) {
        if (deelnemer.nieuwsbriefToestaan) payload.set(naam, veld.value || "on");
      } else if (veld.required) {
        onbekend.push(`verplicht vinkje (${naam})`);
      }
      continue;
    }

    if (veld.type === "radio") {
      if (veld.checked) payload.set(naam, veld.value || "on");
      else if (payload.has(naam)) continue;
      else if (probeerVraag(veld, `${labels[veld.id] || ""} ${naam} ${vraagTekstBij(labels, naam)}`)) continue;
      else if (veld.required) onbekend.push(`verplichte keuze (${naam})`);
      continue;
    }

    const aanwijzing = [veld.name, veld.id, veld.placeholder, labels[veld.id] || ""].join(" ");
    if (probeerVraag(veld, aanwijzing)) continue;

    const sleutel = raadVeld(veld, labels);
    if (sleutel) {
      const waarde = String(deelnemer[sleutel] ?? "").trim();
      // Liever niets versturen dan een leeg verplicht veld: dan weet je meteen
      // welk gegeven nog als secret ontbreekt.
      if (waarde) payload.set(naam, waarde);
      else if (veld.required) onbekend.push(`${sleutel} ontbreekt in de secrets (${naam})`);
    } else if (veld.required) {
      onbekend.push(`verplicht veld (${naam})`);
    }
  }

  return { payload, onbekend, uitleg };
}

// De vraagtekst staat niet altijd in het label van de radio zelf, maar in het
// label van de groep of in een kop erboven; we plakken alles wat we hebben aan
// elkaar zodat de herkenning kans maakt.
function vraagTekstBij(labels, naam) {
  return Object.entries(labels)
    .filter(([id]) => id.toLowerCase().includes(naam.toLowerCase()))
    .map(([, tekst]) => tekst)
    .join(" ");
}

export async function deelnemen(url, deelnemer, opties = {}) {
  const { dryRun = true, context = "" } = opties;

  let html;
  try {
    html = await haal(url);
  } catch (fout) {
    if (fout.name === "BudgetOp") return { status: "later", opmerking: fout.message };
    return { status: "mislukt", opmerking: `pagina niet op te halen: ${fout.message}` };
  }

  if (bevatCaptcha(html)) return { status: "handmatig", opmerking: "captcha op de pagina" };
  const laag = html.toLowerCase();
  if (LOGIN_SPOREN.some((spoor) => laag.includes(spoor))) {
    return { status: "handmatig", opmerking: "lijkt een account of login te vragen" };
  }

  let { formulieren, labels } = leesFormulieren(html);
  let formulier = kiesFormulier(formulieren);
  let doelPagina = url;

  // Geen formulier? Dan staan we waarschijnlijk op de detailpagina van een
  // overzichtssite. Volg de knop naar de echte wedstrijd — één keer.
  if (!formulier) {
    const door = zoekDoorklik(html, url);
    if (!door) return { status: "handmatig", opmerking: "geen deelnameformulier en geen doorklik gevonden" };
    let tweede;
    try {
      tweede = await haal(door);
    } catch (fout) {
      if (fout.name === "BudgetOp") return { status: "later", opmerking: fout.message };
      return { status: "mislukt", opmerking: `doorklik ${door} niet op te halen: ${fout.message}` };
    }
    if (bevatCaptcha(tweede)) {
      return { status: "handmatig", opmerking: `captcha op ${new URL(door).hostname}` };
    }
    ({ formulieren, labels } = leesFormulieren(tweede));
    formulier = kiesFormulier(formulieren);
    doelPagina = door;
    html = tweede;
    if (!formulier) {
      return { status: "handmatig", opmerking: `geen formulier op ${new URL(door).hostname} (waarschijnlijk JavaScript)` };
    }
  }

  const { payload, onbekend, uitleg } = bouwPayload(formulier, labels, deelnemer, {
    pagina: stripHtml(html),
    context,
  });
  const extra = uitleg.length ? ` · ${uitleg.join(" · ")}` : "";
  if (onbekend.length) {
    return { status: "handmatig", opmerking: `niet begrepen: ${onbekend.slice(0, 4).join("; ")}` };
  }
  if (![...payload.keys()].some((naam) => /e.?mail/i.test(naam))) {
    return { status: "handmatig", opmerking: "geen e-mailveld herkend" };
  }

  const doel = new URL(formulier.action || doelPagina, doelPagina).toString();
  if (dryRun) {
    return {
      status: "proefdraai",
      opmerking: `zou versturen naar ${doel} met ${[...payload.keys()].length} velden${extra}`,
    };
  }

  let antwoord;
  try {
    budget.neem();
    antwoord = await fetch(doel, {
      method: "POST",
      body: payload,
      signal: AbortSignal.timeout(25000),
      headers: {
        "User-Agent": UA,
        "Content-Type": "application/x-www-form-urlencoded",
        Referer: doelPagina,
        "Accept-Language": "nl-BE,nl;q=0.9",
      },
    });
  } catch (fout) {
    if (fout.name === "BudgetOp") return { status: "later", opmerking: fout.message };
    return { status: "mislukt", opmerking: `versturen mislukt: ${fout.message}` };
  }

  const resultaat = stripHtml(await antwoord.text()).toLowerCase();
  if (GELUKT.some((markering) => resultaat.includes(markering))) {
    return { status: "gedaan", opmerking: `verstuurd naar ${doel}${extra}` };
  }
  return { status: "handmatig", opmerking: "verstuurd, maar geen bevestiging herkend — zelf nakijken" };
}

File: wedstrijden/deelnemen.py
#!/usr/bin/env python3
"""
Deelnemen: het formulier van een wedstrijd invullen en versturen.

Twee manieren:
  http        → formulier uit de HTML halen en zelf posten (snel, geen browser,
                werkt bij eenvoudige formulieren)
  playwright  → een echte browser laten klikken (voor JavaScript-formulieren;
                enkel als playwright geïnstalleerd is)

En drie uitkomsten:
  gedaan      → verstuurd en de pagina bevestigt het
  proefdraai  → dry_run stond aan; we tonen wat we zouden versturen
  handmatig   → captcha, login, bestand nodig, of geen formulier gevonden;
                de wedstrijd komt in de digest zodat jij kan beslissen

Bewust ingebouwde remmen:
  * dry_run staat standaard aan (config.deelname.dry_run)
  * max_per_nacht en pauze_seconden tussen twee deelnames
  * captcha of een verplicht vinkje dat we niet begrijpen → nooit gokken,
    maar doorschuiven naar "handmatig"
  * per wedstrijd maar één keer (zie opslag.py) — dat is meestal ook de regel
"""
import json
import re
import time
import urllib.parse
import urllib.request
from html.parser import HTMLParser
from pathlib import Path

from bronnen import UA, BronFout, haal, strip_html

CAPTCHA_SPOREN = ("recaptcha", "g-recaptcha", "hcaptcha", "h-captcha", "cf-turnstile",
                  "turnstile", "friendlycaptcha", "captcha")

# Formulieren die géén deelnameformulier zijn. Zonder deze filter kiest de bot
# op elke WordPress-site het reactieformulier (dat heeft ook een e-mailveld) en
# plaatst hij je naam en adres publiek onder een blogartikel.
GEEN_DEELNAME_ACTIE = re.compile(
    r"wp-comments-post|/comment|#comment|/search|/zoek|list-manage\.com|"
    r"mailchimp|newsletter|nieuwsbrief|/login|/inloggen|/account", re.I)

GEEN_DEELNAME_VELDEN = (
    {"comment", "author"},        # WordPress-reacties
    {"comment", "email"},
    {"bericht", "naam"},          # contactformulier
)

ZOEKVELDEN = ({"s"}, {"q"}, {"search"}, {"zoek"}, {"s", "submit"}, {"q", "submit"})

GELUKT_STANDAARD = ("bedankt", "je deelname", "uw deelname", "deelname geregistreerd",
                    "succesvol", "gelukt", "bevestigingsmail", "thank you", "veel succes")

# Hoe we een invoerveld herkennen aan zijn name/id/label.
VELDPATRONEN = [
    (r"voornaam|firstname|first_name|fname|given", "voornaam"),
    (r"achternaam|lastname|last_name|lname|surname|familienaam", "achternaam"),
    (r"\bnaam\b|fullname|full_name|\bname\b", "volledige_naam"),
    (r"e.?mail", "email"),
    (r"telefoon|gsm|mobiel|phone|tel\b", "telefoon"),
    (r"straat|street|adres|address(?!.*mail)", "straat"),
    (r"huisnummer|nummer|house.?number|streetnumber", "huisnummer"),
    (r"postcode|zip|postal", "postcode"),
    (r"gemeente|woonplaats|stad|city|plaats", "gemeente"),
    (r"land|country", "land"),
    (r"geboorte|birth|dob", "geboortedatum"),
]


class FormulierParser(HTMLParser):
    """Haalt de formulieren met hun velden en labels uit een pagina."""

    def __init__(self):
        super().__init__(convert_charrefs=True)
        self.formulieren: list[dict] = []
        self._huidig: dict | None = None
        self._label_voor: str | None = None
        self._label_tekst: list[str] = []
        self.labels: dict[str, str] = {}

    def handle_starttag(self, tag, attrs):
        kenmerken = {k.lower(): (v or "") for k, v in attrs}
        if tag == "form":
            self._huidig = {"action": kenmerken.get("action", ""),
                            "method": (kenmerken.get("method") or "get").lower(),
                            "velden": []}
        elif tag in ("input", "select", "textarea") and self._huidig is not None:
            self._huidig["velden"].append({
                "tag": tag,
                "type": (kenmerken.get("type") or ("select" if tag == "select" else "text")).lower(),
                "name": kenmerken.get("name", ""),
                "id": kenmerken.get("id", ""),
                "value": kenmerken.get("value", ""),
                "placeholder": kenmerken.get("placeholder", ""),
                "required": "required" in kenmerken,
                "checked": "checked" in kenmerken,
            })
        elif tag == "label":
            self._label_voor = kenmerken.get("for", "")
            self._label_tekst = []

    def handle_data(self, data):
        if self._label_voor is not None:
            self._label_tekst.append(data)

    def handle_endtag(self, tag):
        if tag == "form" and self._huidig is not None:
            self.formulieren.append(self._huidig)
            self._huidig = None
        elif tag == "label" and self._label_voor is not None:
            if self._label_voor:
                self.labels[self._label_voor] = " ".join("".join(self._label_tekst).split())
            self._label_voor, self._label_tekst = None, []


# --------------------------------------------------------------------------- recepten

def laad_recepten(map_pad: Path) -> list[dict]:
    """Recepten zijn handmatige uitzonderingen per site; zie recepten/_voorbeeld.json."""
    recepten = []
    for pad in sorted(Path(map_pad).glob("*.json")):
        if pad.name.startswith("_"):
            continue
        try:
            recepten.append(json.loads(pad.read_text(encoding="utf-8")))
        except json.JSONDecodeError as fout:
            print(f"  ! recept {pad.name} is geen geldige JSON: {fout}")
    return recepten


def zoek_recept(url: str, recepten: list[dict]) -> dict | None:
    domein = urllib.parse.urlsplit(url).netloc.lower().removeprefix("www.")
    for recept in recepten:
        if recept.get("url_patroon") and re.search(recept["url_patroon"], url, re.I):
            return recept
        if recept.get("domein", "").lower().removeprefix("www.") == domein:
            return recept
    return None


# --------------------------------------------------------------------------- invullen

def waarden(deelnemer: dict) -> dict:
    """De gegevens die in een formulier terechtkomen."""
    gegevens = dict(deelnemer)
    gegevens["volledige_naam"] = f"{deelnemer.get('voornaam','')} {deelnemer.get('achternaam','')}".strip()
    return gegevens


def vul_sjabloon(sjabloon: str, gegevens: dict) -> str:
    """Vervangt {{voornaam}} en co. in een receptwaarde."""
    def vervang(treffer):
        return str(gegevens.get(treffer.group(1).strip(), ""))
    return re.sub(r"\{\{([^}]+)\}\}", vervang, sjabloon)


def raad_veld(veld: dict, labels: dict) -> str | None:
    """Welk gegeven hoort in dit invoerveld? None = onbekend."""
    if veld["type"] == "email":
        return "email"
    if veld["type"] == "tel":
        return "telefoon"
    aanwijzing = " ".join([veld.get("name", ""), veld.get("id", ""),
                           veld.get("placeholder", ""), labels.get(veld.get("id", ""), "")]).lower()
    for patroon, sleutel in VELDPATRONEN:
        if re.search(patroon, aanwijzing):
            return sleutel
    return None


def is_deelnameformulier(formulier: dict) -> bool:
    """Zeef reactie-, zoek-, nieuwsbrief- en loginformulieren eruit."""
    if GEEN_DEELNAME_ACTIE.search(formulier.get("action", "")):
        return False

    namen = {(v.get("name") or "").lower() for v in formulier["velden"] if v.get("name")}
    if any(kern <= namen for kern in GEEN_DEELNAME_VELDEN):
        return False
    if any(namen == zoek for zoek in ZOEKVELDEN):
        return False

    invulbaar = [v for v in formulier["velden"]
                 if v["type"] not in ("hidden", "submit", "button", "image")]
    # Eén enkel e-mailveld is een nieuwsbriefinschrijving, geen wedstrijd.
    if len(invulbaar) <= 1:
        return False
    return True


def kies_formulier(formulieren: list[dict]) -> dict | None:
    """Het deelnameformulier: heeft een e-mailveld, en is er ook echt één."""
    kandidaten = []
    for formulier in formulieren:
        velden = formulier["velden"]
        heeft_email = any(v["type"] == "email" or re.search(r"e.?mail", v.get("name", ""), re.I) for v in velden)
        zichtbaar = [v for v in velden if v["type"] not in ("hidden", "submit", "button", "image")]
        if heeft_email and zichtbaar and is_deelnameformulier(formulier):
            kandidaten.append((len(zichtbaar), formulier))
    if not kandidaten:
        return None
    return max(kandidaten, key=lambda p: p[0])[1]


def bouw_payload(formulier: dict, labels: dict, gegevens: dict,
                 recept: dict | None) -> tuple[dict, list[str]]:
    """Bouwt de POST-gegevens. Tweede return: velden die we niet begrepen."""
    payload: dict[str, str] = {}
    onbekend: list[str] = []
    vaste_velden = (recept or {}).get("velden", {})

    for veld in formulier["velden"]:
        naam = veld.get("name")
        if not naam or veld["type"] in ("submit", "button", "image", "file", "password"):
            if veld["type"] in ("file", "password"):
                onbekend.append(f"{veld['type']}-veld ({naam or veld.get('id')})")
            continue

        if naam in vaste_velden:
            payload[naam] = vul_sjabloon(str(vaste_velden[naam]), gegevens)
            continue

        if veld["type"] == "hidden":
            payload[naam] = veld.get("value", "")
            continue

        if veld["type"] == "checkbox":
            label = (labels.get(veld.get("id", ""), "") + " " + naam).lower()
            # Voorwaarden en leeftijdsbevestiging aanvinken; reclame enkel als het mag.
            if re.search(r"voorwaard|reglement|akkoord|privacy|18|leeftijd|terms", label):
                payload[naam] = veld.get("value") or "on"
            elif re.search(r"nieuwsbrief|newsletter|aanbieding|partner|marketing|mailing", label):
                if gegevens.get("nieuwsbrief_toestaan"):
                    payload[naam] = veld.get("value") or "on"
            elif veld["required"]:
                onbekend.append(f"verplicht vinkje ({naam})")
            continue

        if veld["type"] == "radio":
            if veld.get("checked"):
                payload[naam] = veld.get("value", "on")
            elif veld["required"] and naam not in payload:
                onbekend.append(f"verplichte keuze ({naam})")
            continue

        sleutel = raad_veld(veld, labels)
        if sleutel:
            waarde = str(gegevens.get(sleutel, "") or "").strip()
            if waarde:
                payload[naam] = waarde
            elif veld["required"]:
                # Liever niets versturen dan een leeg verplicht veld: dan weet je
                # meteen welk gegeven nog in config.json ontbreekt.
                onbekend.append(f"{sleutel} staat niet in config.json ({naam})")
        elif veld["required"]:
            onbekend.append(f"verplicht veld ({naam})")

    return payload, onbekend


def bevat_captcha(html: str) -> bool:
    lage = html.lower()
    return any(spoor in lage for spoor in CAPTCHA_SPOREN)


def gelukt(html: str, recept: dict | None) -> bool:
    tekst = strip_html(html).lower()
    markeringen = (recept or {}).get("gelukt_indien") or GELUKT_STANDAARD
    return any(m.lower() in tekst for m in markeringen)


# --------------------------------------------------------------------------- uitvoeren

def deelnemen_http(url: str, gegevens: dict, recept: dict | None,
                   dry_run: bool, timeout: int) -> tuple[str, str]:
    """Geeft (status, opmerking) terug."""
    try:
        html = haal(url, timeout)
    except BronFout as fout:
        return "mislukt", f"pagina niet op te halen: {fout}"

    if bevat_captcha(html):
        return "handmatig", "captcha op de pagina"

    parser = FormulierParser()
    parser.feed(html)
    formulier = kies_formulier(parser.formulieren)
    if formulier is None:
        return "handmatig", "geen deelnameformulier gevonden op de pagina"

    # Login herkennen we aan een wachtwoordveld ín dit formulier. Op het hele
    # document zoeken sloeg aan op de inlogknop in elke sitenavigatie.
    if any(v["type"] == "password" for v in formulier["velden"]):
        return "handmatig", "het formulier vraagt een account (wachtwoordveld)"

    payload, onbekend = bouw_payload(formulier, parser.labels, gegevens, recept)
    if onbekend:
        return "handmatig", "niet begrepen: " + "; ".join(onbekend[:4])
    if not any(re.search(r"e.?mail", naam, re.I) for naam in payload):
        return "handmatig", "geen e-mailveld herkend"

    doel = urllib.parse.urljoin(url, formulier["action"] or url)
    if dry_run:
        kort = {k: v for k, v in payload.items() if v}
        return "proefdraai", f"zou POST'en naar {doel} met {len(kort)} velden: {', '.join(sorted(kort)[:8])}"

    data = urllib.parse.urlencode(payload).encode("utf-8")
    verzoek = urllib.request.Request(doel, data=data, headers={
        "User-Agent": UA,
        "Content-Type": "application/x-www-form-urlencoded",
        "Referer": url,
        "Accept-Language": "nl-BE,nl;q=0.9",
    })
    try:
        with urllib.request.urlopen(verzoek, timeout=timeout) as antwoord:
            resultaat = antwoord.read().decode("utf-8", errors="replace")
    except Exception as fout:  # netwerk, HTTP-fout, time-out
        return "mislukt", f"versturen mislukt: {fout}"

    if gelukt(resultaat, recept):
        return "gedaan", f"verstuurd naar {doel}"
    return "handmatig", "verstuurd, maar geen bevestiging herkend — zelf nakijken"


def deelnemen_playwright(url: str, gegevens: dict, recept: dict | None,
                         dry_run: bool, config: dict) -> tuple[str, str]:
    """Formulieren die JavaScript nodig hebben. Vereist `pip install playwright`."""
    try:
        from playwright.sync_api import sync_playwright
    except ImportError:
        return "handmatig", "playwright niet geïnstalleerd (pip install playwright)"

    velden = (recept or {}).get("velden", {})
    if not velden:
        return "handmatig", "playwright-modus vraagt een recept met selectors"

    timeout = int(config.get("timeout_seconden", 30)) * 1000
    chromium_pad = config.get("chromium_pad") or None

    with sync_playwright() as pw:
        starter = pw.chromium.launch(headless=True, executable_path=chromium_pad)
        pagina = starter.new_page(locale="nl-BE")
        try:
            pagina.goto(url, timeout=timeout, wait_until="domcontentloaded")
            if bevat_captcha(pagina.content()):
                return "handmatig", "captcha op de pagina"
            for selector, sjabloon in velden.items():
                pagina.fill(selector, vul_sjabloon(str(sjabloon), gegevens), timeout=timeout)
            for selector in (recept or {}).get("aanvinken", []):
                pagina.check(selector, timeout=timeout)
            if dry_run:
                return "proefdraai", f"formulier ingevuld, niet verstuurd ({len(velden)} velden)"
            pagina.click(recept["verzend"], timeout=timeout)
            pagina.wait_for_timeout(3000)
            if gelukt(pagina.content(), recept):
                return "gedaan", "verstuurd via browser"
            return "handmatig", "verstuurd, maar geen bevestiging herkend — zelf nakijken"
        except Exception as fout:
            return "mislukt", f"browser: {type(fout).__name__}: {fout}"
        finally:
            starter.close()


def doe_mee(items: list[dict], config: dict, opslag, log=print) -> list[dict]:
    """Loopt de gekozen wedstrijden af en probeert er aan deel te nemen."""
    instellingen = config.get("deelname", {})
    dry_run = bool(instellingen.get("dry_run", True))
    maximum = int(instellingen.get("max_per_nacht", 8))
    pauze = int(instellingen.get("pauze_seconden", 25))
    timeout = int(instellingen.get("timeout_seconden", 30))
    alleen_met_recept = bool(instellingen.get("alleen_met_recept", True))
    max_pogingen = int(instellingen.get("max_pogingen", 3))

    recepten = laad_recepten(Path(__file__).parent / "recepten")
    gegevens = waarden(config.get("deelnemer", {}))
    resultaten: list[dict] = []
    gedaan = 0

    for item in items:
        if gedaan >= maximum:
            opslag.zet_status(item["url"], "nieuw", "wachtrij vol voor vannacht")
            resultaten.append({**item, "status": "overgeslagen", "opmerking": "max_per_nacht bereikt"})
            continue
        if opslag.al_gedaan(item["url"]):
            continue
        if opslag.pogingen(item["url"]) >= max_pogingen:
            opslag.zet_status(item["url"], "overgeslagen",
                              f"na {max_pogingen} pogingen opgegeven")
            continue

        recept = zoek_recept(item["url"], recepten)
        geprobeerd = True
        if recept is None and alleen_met_recept:
            status, opmerking = "handmatig", "geen recept voor deze site"
            geprobeerd = False  # niets opgehaald: telt niet mee voor max_per_nacht
        elif recept and recept.get("modus") == "playwright":
            status, opmerking = deelnemen_playwright(item["url"], gegevens, recept, dry_run, instellingen)
        else:
            status, opmerking = deelnemen_http(item["url"], gegevens, recept, dry_run, timeout)

        log(f"  [{status}] {item['titel'][:70]} — {opmerking}")
        if geprobeerd:
            opslag.tel_poging(item["url"])
        opslag.zet_status(item["url"], status, opmerking)
        opslag.noteer(item, status, opmerking)
        resultaten.append({**item, "status": status, "opmerking": opmerking})

        if geprobeerd:
            gedaan += 1
            if pauze and gedaan < maximum:
                time.sleep(pauze)

    return resultaten
